woocommerce/woocommerce
# Introduce a product type selection within the new experience

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes woocommerce/woocommerce#41586



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure feature `New product editor` is enabled under `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`
2. Go to `Products` -> `Add new` from the left side menu.
3. Set the `Name` of the product.
4. Under the `General` tab a description `This is a standard product. Change product type` should be shown below the `Basic details` section.
5. When clicking the `Change product type` a dropdown menu should be shown listing all the available product types. 
<img width="686" alt="image" src="https://github.com/woocommerce/woocommerce/assets/13334210/c112df55-7b60-4883-bcfb-731505c16714">

6. From the list the right type should be highlighted in blue as the one corresponding to the type of product you are creating and the whole template of the editor should also matches it.
7. When selecting a different type the product changes must be saved and the editor layout should change according to the selected type.
8. A notice `Product type changed.` should be shown right after.
9. Add some custom product types like follows. You can use a Code Snipped plugin or create a custom plugin as you want.
```php
add_action( 'init', 'register_demo_product_type', 0 );
function register_demo_product_type() {
	class WC_Product_Demo extends WC_Product {
		public function __construct( $product ) {
			parent::__construct( $product );
		}

		public function get_type() {
			return 'demo';
		}
	}
}

add_filter( 'product_type_selector', 'add_demo_product_type' );
function add_demo_product_type( $types ){
	$types[ 'demo' ] = __( 'Demo product', 'dm_product' );
	return $types; 
}
```
11. After refreshing the page the custom types created should be shown in the `More` menu item as a sub menu. 
<img width="729" alt="image" src="https://github.com/woocommerce/woocommerce/assets/13334210/23fc9ae6-fdb7-4449-956b-e6894f767e15">

12. Clicking any of those sub menu items should save all the changes in the editor and redirect you to the classic editor.
13. To register a product template to the added product type add the following.
```php
add_filter( 'woocommerce_product_editor_product_templates', 'register_product_template' );
function register_product_template( $templates ) {
  $templates[] = new Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductTemplate(
    array(
	    'id'                 => 'demo-product-template',
	    'title'              => __( 'Product demo', 'woocommerce' ),
	    'description'        => __( 'This is a demo product', 'woocommerce' ),
	    'order'              => ( count( $templates ) + 1 ) * 10,
	    'icon'               => 'bug',
	    'layout_template_id' => null,
	    'product_data'       => array(
		    'type' => 'demo',
	    ),
    )
    );
	
	return $templates;
}
```
15. After that the menu item should looks like this 
<img width="692" alt="image" src="https://github.com/woocommerce/woocommerce/assets/13334210/37408fe3-5045-417a-8c14-f0e29734562b">

16. To assign an existing layout template just need to set the layout template id to an existing template id `'layout_template_id' => 'simple-product',`
```php
add_filter( 'woocommerce_product_editor_product_templates', 'register_product_template' );
function register_product_template( $templates ) {
	$templates[] = new Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductTemplate(
		array(
			'id'                 => 'demo-product-template',
			'title'              => __( 'Product demo', 'woocommerce' ),
			'description'        => __( 'This is a demo product', 'woocommerce' ),
			'order'              => ( count( $templates ) + 1 ) * 10,
			'icon'               => 'bug',
			'layout_template_id' => 'simple-product',
			'product_data'       => array(
				'type' => 'demo',
			),
		)
	);
	
	return $templates;
}
```
17. With that you added support to the new product type created in point 9. And the menu now should looks like 
<img width="690" alt="image" src="https://github.com/woocommerce/woocommerce/assets/13334210/5f9b138a-a2a5-4631-baf1-cbac9f61a164">

18. Note that the new item now is part of the root dropdown menu and not within the `More` sub menu.
19. Clicking it should save the changes, apply the `simple-product` template to the editor and keep you in the same page without redirecting to the classic experience. (Refreshing the page will still redirect to the classic editor, this is going to be fixed a other PR)



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
